File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution {
    public int minSetSize(int[] arr) {
		
		HashMap<Integer, Integer> map = new HashMap();
		
		for(int nums : arr) {
			map.put(nums, map.getOrDefault(nums, 0) + 1);
		}
		
		int freq[] = new int[map.size()];
		
		int indx = 0;
		for(int key : map.keySet()) {
			freq[indx] = map.get(key);
			indx++;
		}
		
		Arrays.sort(freq);
		System.out.println(Arrays.toString(freq));
		
		int sum = 0, ans = 0;
		int size = arr.length % 2 == 0 ? arr.length : arr.length+1;
		for(int i = freq.length - 1; i >= 0; i--) {
			sum += freq[i];
			ans++;
			if(sum >= size / 2)
				return ans;
		}
		
		return ans;
	}
}

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.py
class Solution:
    def minSetSize(self, arr: List[int]) -> int:
        dic = dict()
        for i in arr:
            if dic.get(i):
                dic[i] += 1
            else:
                dic[i] = 1

        dic_lst = []
        for key in dic:
            dic_lst.append([key, dic.get(key)])
        dic_lst.sort(key=lambda x:x[1], reverse=True)

        count = reduced = 0
        for i, j in dic_lst:
            reduced += j
            count += 1
            if reduced >= len(arr) // 2:
                return count
        return count